pkg/activator/throttler.go

```diff
@@ -0,0 +1,184 @@
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	delete(t.breakers, rev)
+}
+
+// UpdateCapacity update the max concurrency of the Breaker corresponding to the revision
+func (t *Throttler) UpdateCapacity(rev RevisionID, size int32) error {
+	t.mux.RLock()
+	defer t.mux.RUnlock()
+	var containerConcurrency int32
+	serviceName, ok := t.services[rev]
+	if !ok {
+		return fmt.Errorf("no service name registered for the revision: %s", rev)
+	}
+	revision, err := t.getRevision(rev.Namespace, serviceName, t.revisionInformer)
+	if err != nil {
+		return fmt.Errorf("no revision was found for: %s", rev)
+	}
+	containerConcurrency = int32(revision.Spec.ContainerConcurrency)
+	// Equals to the pending queue size as defined in `cmd/queue/main.go`
+	if containerConcurrency < 10 {
+		containerConcurrency = 10
+	}
+	if breaker, ok := t.breakers[rev]; ok {
+		err := breaker.UpdateConcurrency(size * containerConcurrency)
+		if err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("no breaker was found for the received revision: %s", rev)
+	}
+	return nil
+}
+
+// Try executes the `function` on the Breaker
+// Returns true if the function was executed, false otherwise
+// The function is not executed either if the Breaker doesn't exist or it is full
+func (t *Throttler) Try(rev RevisionID, function func()) error {
+	t.mux.RLock()
+	if breaker, ok := t.breakers[rev]; ok {
+		t.mux.RUnlock()
+		ok := breaker.Maybe(function)
+		if !ok {
+			return errors.New("activator throttler is in overflow")
+		}
+		return nil
+	}
+	t.mux.RUnlock()
+	return fmt.Errorf("breaker is not found for the revision %s", rev)
+}
+
+func (t *Throttler) getRevision(namespace, name string, informer v1alpha1.RevisionInformer) (*v1alpha12.Revision, error) {
+	var revision *v1alpha12.Revision
+	revisionKey := cache.ExplicitKey(namespace + "/" + name)
+	rev, exist, err := informer.Informer().GetIndexer().Get(revisionKey)
+	if !exist {
+		return nil, errors.New("revision doesn't exist yet")
+	}
+	if err != nil {
+		return nil, err
+	}
+	revision = rev.(*v1alpha12.Revision)
+	return revision, nil
+}
+
+// Helper defines methods handler methods for endpoints observer
+type Helper struct {
+}
+
+// UpdateEndpoints communicates the new number of endpoints to the Throttler
+func (h Helper) UpdateEndpoints(throttler *Throttler) func(oldObj, newObj interface{}) {
+	return func(oldObj, newObj interface{}) {
+		oldEndpoints := oldObj.(*corev1.Endpoints)
+		newEndpoints := newObj.(*corev1.Endpoints)
+		if len(oldEndpoints.Subsets) == 0 && len(newEndpoints.Subsets) == 0 {
+			return
+		}
+		oldAddresses := h.getEndpointAddresses(oldEndpoints.Subsets)
+		newAddresses := h.getEndpointAddresses(newEndpoints.Subsets)
+		if oldAddresses == newAddresses {
+			return
+		}
+		revID := RevisionID{newEndpoints.Namespace, newEndpoints.Name}
+		// only update capacity of registered revisions
+		if throttler.Exists(revID) {
+			err := throttler.UpdateCapacity(revID, int32(newAddresses-oldAddresses))
```

Comment:
This Exists/Update bundle is racy. Can we just call UpdateCapacity and handle the error if the revision doesn't exist?

Reply:
do we have a situation when the endpoints are updated before the request gets into the activator?
I'm hesitant of handling the error inside - it will produce some logging noise for the endpoints which do not belong to the revisions, e.g. some management pod got restarted